src/libpmemfile-posix/inode.h

```diff
@@ -0,0 +1,228 @@
+ */
+
+#ifndef PMEMFILE_INODE_H
+#define PMEMFILE_INODE_H
+
+#include <stdint.h>
+#include <time.h>
+
+#include "libpmemfile-posix.h"
+#include "layout.h"
+#include "os_thread.h"
+
+/* volatile inode */
+struct pmemfile_vinode {
+	/* reference counter */
+	uint32_t ref;
+
+	/* read-write lock, also protects inode read/writes */
+	os_rwlock_t rwlock;
+
+	/* persistent inode */
+	struct pmemfile_inode *inode;
+
+	/* persistent inode oid */
+	TOID(struct pmemfile_inode) tinode;
+
+#ifdef DEBUG
+	/*
+	 * One of the full paths inode can be reached from.
+	 * Used only for debugging.
+	 */
+	char *path;
+#endif
+
+	/* parent directory, valid only for directories */
+	struct pmemfile_vinode *parent;
+
+	/* pointer to the array of opened inodes */
+	struct {
+		struct pmemfile_inode_array *arr;
+		unsigned idx;
+	} orphaned;
+
+	/* first free block slot */
+	struct block_info {
+		struct pmemfile_block_array *arr;
+		uint32_t idx;
+	} first_free_block;
+
+	/* first used block */
+	struct pmemfile_block *first_block;
+
+	/* tree mapping offsets to blocks */
+	struct ctree *blocks;
+
+	/* space for volatile snapshots */
+	struct {
+		struct block_info first_free_block;
+		struct pmemfile_block *first_block;
+	} snapshot;
+};
+
+/* inode permission information */
+struct inode_perms {
+	/* file flags (contains mode) */
+	uint64_t flags;
+
+	/* owner */
+	uint32_t uid;
+
+	/* group */
+	uint32_t gid;
+};
+
+static inline struct inode_perms
+_vinode_get_perms(struct pmemfile_vinode *vinode)
+{
+	struct inode_perms perms;
+	struct pmemfile_inode *inode = vinode->inode;
+	perms.flags = inode->flags;
+	perms.uid = inode->uid;
+	perms.gid = inode->gid;
+	return perms;
+}
+
+static inline struct inode_perms
+vinode_get_perms(struct pmemfile_vinode *vinode)
+{
+	struct inode_perms perms;
+	os_rwlock_rdlock(&vinode->rwlock);
+	perms = _vinode_get_perms(vinode);
+	os_rwlock_unlock(&vinode->rwlock);
+	return perms;
+}
+
+static inline bool inode_is_dir(const struct pmemfile_inode *inode)
+{
+	return PMEMFILE_S_ISDIR(inode->flags);
+}
+
+static inline bool vinode_is_dir(struct pmemfile_vinode *vinode)
+{
+	return inode_is_dir(vinode->inode);
+}
+
+static inline bool inode_is_regular_file(const struct pmemfile_inode *inode)
+{
+	return PMEMFILE_S_ISREG(inode->flags);
+}
+
+static inline bool vinode_is_regular_file(struct pmemfile_vinode *vinode)
+{
+	return inode_is_regular_file(vinode->inode);
+}
+
+static inline bool inode_is_symlink(const struct pmemfile_inode *inode)
+{
+	return PMEMFILE_S_ISLNK(inode->flags);
+}
+
+static inline bool vinode_is_symlink(struct pmemfile_vinode *vinode)
+{
+	return inode_is_symlink(vinode->inode);
+}
+
+void file_get_time(struct pmemfile_time *t);
+
+struct pmemfile_vinode *inode_alloc(PMEMfilepool *pfp,
+		uint64_t flags,
+		struct pmemfile_vinode *parent,
+		volatile bool *parent_refed,
+		const char *name,
+		size_t namelen);
+
+void inode_free(PMEMfilepool *pfp, TOID(struct pmemfile_inode) tinode);
```

Comment:
This method should should have a different name. The inode_alloc returns a vinode, but this take a TOID ;)
I kinda understand why this is here, I'm just saying that it was confusing at first.